xtide-2.15.5/tide.1
.TH tide 1 "2020-06-27"  "XTide 2"
.SH NAME
tide \- Harmonic tide clock and tide predictor (command line client)
.SH SYNOPSIS
.B tide
[\fB-b ``YYYY-MM-DD HH:MM''\fP]
[\fB-cf ``format-string''\fP]
[\fB-ch pixels\fP]
[\fB-cw pixels\fP]
[\fB-dc color\fP]
[\fB-Dc color\fP]
[\fB-df ``format-string''\fP]
[\fB-e ``YYYY-MM-DD HH:MM''\fP]
[\fB-ec color\fP]
[\fB-el y|n\fP]
[\fB-em pSsMm\fP]
[\fB-f c|h|i|l|p|t|v\fP]
[\fB-fc color\fP]
[\fB-fg color\fP]
[\fB-ga aspect\fP]
[\fB-gh pixels\fP]
[\fB-gs d|l|s\fP]
[\fB-gt y|n\fP]
[\fB-gw pixels\fP]
[\fB-hf ``format-string''\fP]
[\fB-in y|n\fP]
[\fB-l ``location name''\fP]
[\fB-lb y|n\fP]
[\fB-lw pixels\fP]
[\fB-m a|b|c|C|g|k|l|m|p|r|s\fP]
[\fB-mc color\fP]
[\fB-Mc color\fP]
[\fB-ml [-]N.NN(ft|m|kt)\fP]
[\fB-nc color\fP]
[\fB-o ``filename''\fP]
[\fB-ou y|n\fP]
[\fB-pb y|n\fP]
[\fB-ph mm\fP]
[\fB-pi days\fP]
[\fB-pm mm\fP]
[\fB-pw mm\fP]
[\fB-s ``HH:MM''\fP]
[\fB-tf ``format-string''\fP]
[\fB-th characters\fP]
[\fB-tl y|n\fP]
[\fB-to opacity\fP]
[\fB-tw characters\fP]
[\fB-u ft|m\fP]
[\fB-v\fP]
[\fB-z y|n\fP]
.SH DESCRIPTION
XTide is a package that provides tide and current predictions in a wide
variety of formats. Graphs, text listings, and calendars can be generated,
or a tide clock can be provided on your desktop.
.LP
XTide can work with X-windows, plain text terminals, or the web. This is
accomplished with three separate programs: the interactive interface
(\fBxtide\fP), the non-interactive or command line interface
(\fBtide\fP), and the web interface (\fBxttpd\fP).
.LP
.B tide
is the non-interactive command line client for XTide 2.
For a complete
discussion, please see the verbose documentation at
.BR https://flaterco.com/xtide/ .
.SH CONFIGURATION
Unless a configuration file
.B /etc/xtide.conf
is supplied, you must set the environment variable
.B HFILE_PATH
to point to the harmonics
files that should have been installed along with the
.B xtide
program.  Example:
.RS
export HFILE_PATH=/usr/local/share/xtide/harmonics.tcd
.RE
If a configuration file is used, the first line should consist of the
colon-separated list of harmonics files (e.g. the
.B HFILE_PATH
line above without the "export HFILE_PATH=" part).
The environment variable takes precedence over the configuration file.
.LP
Set the environment variable
.B XTIDE_DEFAULT_LOCATION
to the name of your favorite place if you want to skip the location-choosing
step.
.SH OPTIONS
.TP
.B -b ``YYYY-MM-DD HH:MM''
Specify the begin (start) time for predictions using the ISO 8601
compliant format YYYY-MM-DD HH:MM, where hours
range from 00 to 23.  The timestamp is in the local time
zone for the location, or in UTC if the
.B -z
setting is engaged.  If clock mode is selected or if no
.B -b
is supplied, the current time will be used.
.TP
.B -cf ``format-string''
.BR strftime (3)
style format string for printing days in calendars (default %a %d).
.TP
.B -ch pixels
Height for tide clocks (default 312 pixels).
.TP
.B -cw pixels
Width for tide clocks (default 84 pixels).
.TP
.B -dc color
Daytime background color in tide graphs (default SkyBlue).
.TP
.B -Dc color
Color of datum line in tide graphs (default white).
.TP
.B -df ``format-string''
.BR strftime (3)
style format string for printing dates (default %Y-%m-%d).  For calendars see
.BR -cf .
.TP
.B -e ``YYYY-MM-DD HH:MM''
Specify the end (stop) time for predictions in banner, calendar, alt. calendar,
medium rare, plain, raw, or stats mode.  Does not apply in graph and clock
modes:  the effective end time for graph and clock modes is entirely determined
by the start time
.RB ( -b ),
the width
.RB ( -cw ,
.B -gw
or
.BR -tw ),
and the aspect
.RB ( -ga ).
The format and time zone are the same as for
.BR -b .
If no
.B -e
is supplied, the end time will be set to four days after the begin time.
.TP
.B -ec color
Foreground in tide graphs during outgoing tide (default SeaGreen).
.TP
.B -el y|n
Select whether or not to draw datum and middle-level lines in tide graphs
(default n).
.IP
The National Ocean Service (NOS) defines both Mean Sea Level (MSL) and Mean
Tide Level (MTL) in terms of averages taken over observations.  The
middle-level line is drawn at the midpoint of the theoretical tidal range,
which usually corresponds to the mathematical mean level of the predictions.
This approximates both MSL and MTL, but, strictly speaking, is equivalent to
neither.  Moreover, subordinate station offsets may shift the actual mean so
that it no longer falls at the midpoint of the tidal range.  The datum line
is drawn at the zero level of the predictions, which usually corresponds to
the station's benchmark, but this too can be rendered inaccurate by
subordinate station offsets.
.TP
.B -em pSsMm
Events to suppress (p = phase of moon, S = sunrise, s = sunset, M = moonrise, m
= moonset), or x to suppress none.  E.g, to suppress all sun and moon events,
set eventmask to the value pSsMm.
.TP
.B -f c|h|i|l|p|t|v
Specify the output format as CSV, HTML, iCalendar, LaTeX, PNG, text, or SVG.  The default is text.
.TP
.B -fc color
Foreground in tide graphs during incoming tide (default Blue).
.TP
.B -fg color
Color of text and other notations (default black).
.TP
.B -ga aspect
Aspect for tide graphs, clocks, and banners (default 1.0).  This
controls not the width or height of the window but rather the relative scaling
of the X and Y axes in the plot of the tide curve.
.TP
.B -gh pixels
Height for tide graphs (default 312 pixels).
.TP
.B -gs d|l|s
Style of graphs and clocks.  Valid values: d (default), l (line), s
(semitransparent).
.TP
.B -gt y|n
Select whether or not to label tenths of units in tide graphs
(default n).
.TP
.B -gw pixels
Width for tide graphs (default 960 pixels).
.TP
.B -hf ``format-string''
.BR strftime (3)
style format string for printing hour labels on time axis (default %l).
.TP
.B -in y|n
Select whether or not to use inferred values for some constituents (default n).
For expert use only.
.TP
.B -l ``location name''
Specify a location for tide predictions.  You get the first station where the
name supplied with -l is a case-insensitive match with the beginning
(or the entirety) of the station's name.
.TP
.B -lb y|n
Select whether or not calendar mode inserts a line break before prediction
values (default n).
.TP
.B -lw pixels
Width of line in graph styles l and s
(default 2.5 pixels).
.TP
.B -m a|b|c|C|g|k|l|m|p|r|s
Specify mode to be about, banner, calendar, alt. calendar, graph, clock, list, medium rare, plain, raw, or
stats.  The default is plain.
.TP
.B -mc color
Color of mark line in graphs (default red).
.TP
.B -Mc color
Color of middle-level line in tide graphs (default yellow).
.TP
.B -ml [-]N.NN(ft|m|kt)
Specify the mark level to be used in predictions.  The predictions will include
the times when the tide level crosses the mark.  Not supported in clock
mode.  Example usage: -ml -0.25ft
.TP
.B -nc color
Nighttime background color in tide graphs (default DeepSkyBlue).
.TP
.B -o ``filename''
Redirect output to the specified file (appends).
.TP
.B -ou y|n
Select whether or not to print numbers with no ft/m/kt
(default n). Where possible, adds a header line stating the units and datum.
.TP
.B -pb y|n
Select whether or not calendar mode inserts a page
break and repeats the station header for each month in text, HTML, and LaTeX
formats (default y).
.TP
.B -ph mm
Nominal length of paper in LaTeX output (default 420 mm).  This need not match
your actual paper; use "Shrink oversized pages" in print options.
.TP
.B -pi days
Number of days of predictions to generate when no end time is specified (default 4).
.TP
.B -pm mm
Nominal width of top, bottom, left and right margins in LaTeX output (default
10 mm).  Actual width will depend on print scaling.
.TP
.B -pw mm
Nominal width of paper in LaTeX output (default 297 mm). This need not match your actual paper; use "Shrink oversized pages" in print options.
.TP
.B -s ``HH:MM''
Specify the step interval, in hours and minutes, for raw
or medium rare mode predictions.  The default is one hour.
.TP
.B -tf ``format-string''
.BR strftime (3)
style format string for printing times (default %l:%M %p %Z).
.TP
.B -th characters
Height of ASCII graphs and clocks (default 24 characters).
.TP
.B -tl y|n
Select whether or not to draw depth lines on top of the tide graph
(default n).
.TP
.B -to opacity
Opacity of the fill in graph style s (semitransparent) (default 0.65).
.TP
.B -tw characters
Width of text format (default 79 characters).
.TP
.B -u ft|m
Preferred units of length (default no preference).
.TP
.B -v
Print version string and exit.
.TP
.B -z y|n
Select whether or not to coerce all time zones to UTC (default n).
.SH NOTES
If the environment variable
.B TERM
is set to vt100 or vt102,
.B tide
will take it seriously and invoke the DEC Special Graphics character set in
text form output.
.SH "SEE ALSO"
.BR xtide (1),
.BR xttpd (8),
.BR https://flaterco.com/xtide/ .
.sp


xtide-2.15.5/NEWS
   [1]<- Previous [2]Contents

Icon News

XTide 2.15.5 (libxtide.so.1.0.4) 2022-05-11

   XTide 2.15.5 restored compatibility with recent gpsd.  XTide 2.15.4
   fixed a FTB with libXaw 1.0.14 and assorted minor issues.  Get the
   latest from [3]https://flaterco.com/xtide/files.html and see the
   [4]changelog for details.

Data refresh 2022-01-09

   The data refresh available from
   [5]https://flaterco.com/xtide/files.html#harmonicsfiles includes new
   tide stations for Puerto Rico and the Virgin Islands.

   A corresponding update of the harmonic constants management package is
   available from [6]https://flaterco.com/xtide/files.html#experts.

ftp is dead

   The archive is at [7]https://flaterco.com/files/xtide/.  Old links
   pointing to ftp://ftp.flaterco.com/xtide/... should be redirected to
   https://flaterco.com/files/xtide/....

End of maintenance for non-free data

   For [8]various reasons, I have ended maintenance of the non-free data
   set.  Anyone interested in generating new harmonic constants from
   published data can find information and a list of data sources [9]here.

Mailing list

   There is an announce-only mailing list to which I send notices of each
   update to XTide or the harmonics data.  You can subscribe or read the
   archives at
   [10]https://pairlist10.pair.net/mailman/listinfo/xtide-announce.

   Please note:  If an announcement to you gets bounced by your mail
   server, you fall off the list.  Some long-time contributors have fallen
   off because, unbeknownst to them, their ISPs subscribed to a spam
   blackholing service that blackholed all pair.com-hosted domains after
   one of them sent out spam.

To do list

   XTide
     * Support [11]QGIS integration as a substitute for the builtin
       location chooser.
     * Support m/s as a unit for currents.  NOAA default is now cm/s.
     * If libxtide is changed, consider merging some accommodations from
       XTideMac.
          + Opting out of LibPNG, RGBGraph, Settings, and codeset support.
          + Compiler was opinionated about character encoding and int
            types.
          + Configure for dirent.h, syslog.h.
     * Purge antique portability workarounds.
     * Make xttpd use https.
     * Support lookup by station ID instead of only name.
     * The XML parser should be bundled under the libxtide namespace but
       isn't.

   Major things to do in XTide 3 if life turns out not to be too short
   after all:
     * Support minor axes for currents; yield speed and direction at any
       time; plot as a cylinder.
     * Find a way to put an end to the infinite number of customization
       settings for graph mode, such as by dumping ALL of the data,
       metadata, and events to a single-file format that an external
       graphing app will read.
     * Renovate programming practices again (exceptions, checked and const
       operations, strings).
     * On exceptions:  get the application-specific error messages OUT of
       libxtide.
     * Move from Latin-1 to UTF-8 (requires libtcd revision).
     * Harmonize the globe location chooser with the way that Google Earth
       operates.  Use OpenGL of course.
     * Support Wayland in non-X11 mode.
     * Hydraulic current support is obsolete.

   harmbase2

   All the cheap and nasty reverse geocoding for the purpose of improving
   station descriptions is stupid and should be punted to [12]QGIS.

   harmgen

   It should be possible to [13]link directly with liboctave to avoid the
   very inefficient conversion of input data from binary to text and back
   again that currently is done.  However, this would be fixing something
   that is not broken, and is neither important nor urgent.

   libtcd

   Purge antique portability workarounds from configure.ac.  Always just
   include stdint not inttypes.

   In signed_bit_unpack, extend_mask << numbits uses undefined behavior.

   In the documentation section on inference, cite NOAA SP NOS CO-OPS 3
   Sec. 4.2.3 explaining how bogus it is.

   Things to do in the next major revision, if there ever is one:
     * Purge COMPAT114 code and documentation.
     * UTF-8.
     * Move station ID to header.
     * Add header field to flag tides versus currents.
     * Add depth field for currents.
     * Support minor axes for currents.
     * Get rid of internal state; fix inability to have more than one
       database open.
     * Remove gratuitous complexity in the encoding of speeds, equilibrium
       arguments, and node factors; make speeds 32-bit unsigned with no
       offset.
     * Possibly permit node factors to be specified for intervals shorter
       than 1 year and/or support apps that use libcongen to generate the
       node factors themselves.
     __________________________________________________________________

   [14]<- Previous [15]Contents

References

   1. https://flaterco.com/xtide/changelog.html
   2. https://flaterco.com/xtide/xtide.html#contents
   3. https://flaterco.com/xtide/files.html
   4. https://flaterco.com/xtide/changelog.html
   5. https://flaterco.com/xtide/files.html#harmonicsfiles
   6. https://flaterco.com/xtide/files.html#experts
   7. https://flaterco.com/files/xtide/
   8. https://flaterco.com/xtide/faq.html#renaissance
   9. https://flaterco.com/xtide/harmonics.html#harmgen
  10. https://pairlist10.pair.net/mailman/listinfo/xtide-announce
  11. https://qgis.org/en/site/
  12. https://qgis.org/en/site/
  13. https://www.gnu.org/software/octave/doc/interpreter/Standalone-Programs.html
  14. https://flaterco.com/xtide/changelog.html
  15. https://flaterco.com/xtide/xtide.html#contents


xtide-2.15.5/tide.cc
// $Id: tide.cc 7306 2020-06-28 00:24:42Z flaterco $

/*  tide  Command-line client for dusty old TTYs and line printers.

    Copyright (C) 1998  David Flater.

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "libxtide/libxtide.hh"
using namespace libxtide;


// Separator between tide stations in text client output
// \f is form feed
static constString stationSeparator = "\f";


static void doLocation (const Dstr &name,
                        Interval step,
                        FILE *outfp,
                        Mode::Mode mode,
                        Format::Format form) {
  const StationRef *sr (Global::stationIndex().getStationRefByName(name));
  if (sr) {
    std::auto_ptr<Station> station (sr->load());
    station->step = step;

    // Install mark level, if applicable.
    {
      Configurable &cfbl = Global::settings["ml"];
      if (!cfbl.isNull) {
        station->markLevel = cfbl.p;
        if (cfbl.p.Units() != station->predictUnits())
          station->markLevel.Units (station->predictUnits());
      }
    }

    // Find start and end times.
    // This needs to be done here due to the timezones.
    Timestamp startTime, endTime;
    {
      Configurable &cfbl = Global::settings["b"];
      if (cfbl.isNull)
	startTime = (time_t)time(NULL);
      else {
	startTime = Timestamp (cfbl.s, sr->timezone);
	if (startTime.isNull())
	  Global::cant_mktime (cfbl.s, sr->timezone, Error::fatal);
      }
    }{
      Configurable &ecfg = Global::settings["e"];
      if (ecfg.isNull) {
        Configurable &picfg = Global::settings["pi"];
	endTime = startTime + Interval(picfg.u*DAYSECONDS);
      } else {
	endTime = Timestamp (ecfg.s, sr->timezone);
	if (endTime.isNull())
	  Global::cant_mktime (ecfg.s, sr->timezone, Error::fatal);
      }
    }
    if (startTime > endTime) {
      static bool I_told_you_once (false);
      if (!I_told_you_once) {
        fprintf (stderr, "Tide Warning:  swapping begin and end times\n");
        I_told_you_once = true;
      }
      std::swap (startTime, endTime);
    }

    if (mode == Mode::graph && form == Format::PNG)
      station->graphModePNG (outfp, startTime);
    else if (mode == Mode::clock && form == Format::PNG)
      station->clockModePNG (outfp);
    else {
      Dstr text_out;
      station->print (text_out, startTime, endTime, mode, form);
      fprintf (outfp, "%s", text_out.aschar());
    }

  } else {
    Dstr details ("Could not find: ");
    details += name;
    Global::barf (Error::STATION_NOT_FOUND, details, Error::nonfatal);
  }
}


static void loopLocations (FILE *outfp) {
  bool isFirst (true);

  Interval step (Global::hour);
  {
    Configurable &cfbl = Global::settings["s"];
    if (!cfbl.isNull) {
      step = Interval (cfbl.s);
      if (step <= Global::zeroInterval) {
	Dstr details ("You must specify a positive step.  You tried ");
	details += cfbl.s;
	Global::barf (Error::NUMBER_RANGE_ERROR, details);
      }
    }
  }

  Mode::Mode mode;
  Format::Format form;
  {
    Configurable &cfbl = Global::settings["m"];
    if (cfbl.isNull)
      mode = Mode::plain;
    else
      mode = (Mode::Mode)cfbl.c;
  }{
    Configurable &cfbl = Global::settings["f"];
    if (cfbl.isNull)
      form = Format::text;
    else
      form = (Format::Format)cfbl.c;
  }

  if (mode != Mode::list) {
    Configurable &cfbl = Global::settings["l"];
    DstrVector::const_iterator it = cfbl.v.begin();
    DstrVector::const_iterator stop = cfbl.v.end();
    if (it == stop)
      fprintf (stderr, "Warning:  No locations specified with -l; hence, no output.\n");
    while (it != stop) {
      if (isFirst)
	isFirst = false;
      else
	fprintf (outfp, "%s", stationSeparator);
      doLocation (*it, step, outfp, mode, form);
      ++it;
    }
  } else {
    // List mode
    Dstr text_out;
    Global::stationIndex().print (text_out, form);
    fprintf (outfp, "%s", text_out.aschar());
  }
}


int main (int argc, char **argv) {
  srand (time (NULL));
  Global::initTerm();
  Global::settings.applyUserDefaults();
  Global::settings.applyCommandLine (argc, argv);
  Global::settings.fixUpDeprecatedSettings();

  if (!Global::disclaimerDisabled()) {
    fprintf (stderr, "\
-----------------------------------------------------------------------------\n\
            XTide 2   Copyright (C) 1998 David Flater.\n\
\n\
This software is provided under the terms of the GNU General Public\n\
License, either version 3 of the License, or (at your option) any later\n\
version.\n\
\n\
Although the package as a whole is GPL, some individual source files\n\
are public domain.  Consult their header comments for details.\n\
\n\
                        NOT FOR NAVIGATION\n\
\n\
This program is distributed in the hope that it will be useful, but\n\
WITHOUT ANY WARRANTY; without even the implied warranty of\n\
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  The author\n\
assumes no liability for damages arising from use of this program OR\n\
of any 'harmonics data' that might be distributed with it.  For details, see\n\
the verbose documentation at https://flaterco.com/xtide/.\n\
\n\
This obnoxious message will go away permanently if you click \"Don't show\n\
this again\" in the disclaimer window of the X windows client (xtide), or\n\
if you create a file in your home directory called \".disableXTidedisclaimer\".\n\
-----------------------------------------------------------------------------\n\
\n");
  }

  if (argc < 2 && Global::settings["l"].v.empty()) {
    fprintf (stderr, "\
Minimal usage:  tide -l \"Location name\" (or set the environment variable\n\
  XTIDE_DEFAULT_LOCATION to \"Location name\")\n\n\
Other switches:\n\
  -b \"YYYY-MM-DD HH:MM\"\n\
      Specify the begin (start) time for predictions.\n\
  -e \"YYYY-MM-DD HH:MM\"\n\
      Specify the end (stop) time for predictions.\n\
  -f c|h|i|l|p|t|v\n\
      Specify the output format as CSV, HTML, iCalendar, LaTeX, PNG, text, or\n\
      SVG.  The default is text.\n\
  -m a|b|c|C|g|k|l|m|p|r|s\n\
      Specify mode to be about, banner, calendar, alt. calendar, graph, clock,\n\
      list, medium rare, plain, raw, or stats.  The default is plain.\n\
  -o \"filename\"\n\
      Redirect output to the specified file (appends).\n\
  -s \"HH:MM\"\n\
      Specify the step interval, in hours and minutes, for raw\n\
      mode predictions.  The default is one hour.\n\
  -v\n\
      Print version string and exit.\n\
\n\
These are only the most important switches.  For information on all of\n\
the switches, please read the verbose documentation at:\n\
  https://flaterco.com/xtide/\n");
    exit (-1);
  }

  FILE *outfp = stdout;
  {
    Configurable &cfbl = Global::settings["o"];
    if (!cfbl.isNull)
      if (!(outfp = fopen (cfbl.s.aschar(), "ab")))
	Global::cantOpenFile (cfbl.s, Error::fatal);
  }

  #ifdef NO_INITIALIZER_LISTS
  Global::initializeGraphFont();
  #endif

  loopLocations (outfp);

  exit (0);
}

// Cleanup2006 Done
